ddht/v5_1/alexandria/advertisement_manager.py

```diff
@@ -0,0 +1,257 @@
+from ddht.v5_1.alexandria.messages import AdvertiseMessage
+from ddht.v5_1.alexandria.payloads import Advertisement
+from ddht.v5_1.alexandria.sedes import content_sedes
+
+
+class AdvertisementManager(Service, AdvertisementManagerAPI):
+    logger = logging.getLogger("ddht.AdvertisementManager")
+
+    def __init__(
+        self, network: AlexandriaNetworkAPI, advertisement_db: AdvertisementDatabaseAPI,
+    ) -> None:
+        self._network = network
+        self._advertisement_db = advertisement_db
+        self._ready = trio.Event()
+
+    async def run(self) -> None:
+        self.manager.run_daemon_task(self._handle_advertisement_requests)
+
+        await self.manager.wait_finished()
+
+    async def ready(self) -> None:
+        await self._ready.wait()
+
+    async def _handle_advertisement_requests(self) -> None:
+        async with trio.open_nursery() as nursery:
+            async with self._network.client.subscribe(AdvertiseMessage) as subscription:
+                self._ready.set()
+                async for request in subscription:
+                    if not request.message.payload:
+                        continue
+
+                    nursery.start_soon(self._handle_request, request)
+
+    def check_interest(self, advertisement: Advertisement) -> bool:
+        if self._advertisement_db.exists(advertisement):
+            return False
+
+        distance_from_content = compute_content_distance(
+            self._network.local_node_id, advertisement.content_id,
+        )
+        if distance_from_content > self._network.local_advertisement_radius:
+            return False
+
+        return True
+
+    async def validate_advertisement(self, advertisement: Advertisement) -> None:
+        if not advertisement.is_valid:
+            raise ValidationError(
+                f"Advertisement has invalid signature: signature={advertisement.signature}"
+            )
+
+        if advertisement.is_expired:
+            raise ValidationError(
+                f"Advertisement expired: advertisement={advertisement}"
+            )
+
+        if self._advertisement_db.exists(advertisement):
+            raise ValidationError(
+                f"Advertisement already known: advertisement={advertisement}"
+            )
+
+        # Verify that the `hash_tree_root` matches the known roots.
+        known_hash_tree_roots = set(
+            self._advertisement_db.get_hash_tree_roots_for_content_id(
+                advertisement.content_id
+            )
+        )
+        if len(known_hash_tree_roots) > 1:
+            self.logger.error(
+                "Multiple hash tree roots found: content_id=%s  roots=%s",
+                advertisement.content_id.hex(),
+                known_hash_tree_roots,
+            )
+
+        if (
+            known_hash_tree_roots
+            and advertisement.hash_tree_root not in known_hash_tree_roots
+        ):
+            # TODO: perform full validation of the content to determine the
```

Comment:
This TODO will be addressed later as its a situation that should only occur in adversarial situations and until we have the content management built its difficult to write this logic.